Implement Shopping Cart logic:

DEz-1996 opened the pull request:
Add ShoppingCart and CartItem classes;
Add Dto: CartItemAddBookRequestDto, CartItemQuantityRequestDto, ShoppingCartCreateRequestDto, CartItemResponseDto, ShoppingCartResponseDto, ShoppingCartUpdateRequestDto; Add ShoppingCartMapper and CartItemMapper;
Add changes to create tables "shopping_carts" and "cart_items" in liquibase; Add ShoppingCartService and ShoppingCartServiceImpl; Add ShoppingCartRepository and ShoppingCartRepository; Add ShoppingCartController;

DEz-1996 committed:
Implement Shopping Cart logic:
Add ShoppingCart and CartItem classes;
Add Dto: CartItemAddBookRequestDto, CartItemQuantityRequestDto, ShoppingCartCreateRequestDto, CartItemResponseDto, ShoppingCartResponseDto, ShoppingCartUpdateRequestDto;
Add ShoppingCartMapper and CartItemMapper;
Add changes to create tables "shopping_carts" and "cart_items" in liquibase;
Add ShoppingCartService and ShoppingCartServiceImpl;
Add ShoppingCartRepository and ShoppingCartRepository;
Add ShoppingCartController;

vadymhrnk commented on src/main/java/online/book/store/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,70 @@
+    @PostMapping
+    @Operation(summary = "Add book to shopping cart", description = "Add book to shopping cart")
+    public ShoppingCartResponseDto addBookToShoppingCart(
+            @RequestBody CartItemAddBookRequestDto requestDto,
```
I think it's better to add validation for this DTO

vadymhrnk commented on src/main/java/online/book/store/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,70 @@
+    @Operation(summary = "Update quantity of a book in the shopping cart",
+            description = "Update quantity of a book in the shopping cart")
+    public ShoppingCartResponseDto updateItemsQuantity(
+                                       @RequestBody CartItemQuantityRequestDto quantityRequestDto,
```
Same here

vadymhrnk commented on src/main/java/online/book/store/dto/ShoppingCartCreateRequestDto.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Data;
+
+@Data
+public class ShoppingCartCreateRequestDto {
```
I think this DTO is redundant, because I can't find where you use it

vadymhrnk commented on src/main/java/online/book/store/dto/ShoppingCartUpdateRequestDto.java:
```diff
@@ -0,0 +1,8 @@
+import lombok.Data;
+
+@Data
+public class ShoppingCartUpdateRequestDto {
```
Same here, redundant DTO

vadymhrnk commented on src/main/java/online/book/store/mapper/ShoppingCartMapper.java:
```diff
@@ -0,0 +1,14 @@
+
+@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
+public interface ShoppingCartMapper {
+
```
Redundant empty line

vadymhrnk commented on src/main/java/online/book/store/model/CartItem.java:
```diff
@@ -0,0 +1,43 @@
+@Data
+@Entity
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
```
It's up to you, but you can use `@SQLRestriction("is_deleted=false")` instead of deprecated `@Where`

d1sam replied:
agree :)

DEz-1996 replied:
> agree :)

Yes, I added this to the list of fixing commit. Because I used @Where everywhere in my code

vadymhrnk commented on src/main/java/online/book/store/model/ShoppingCart.java:
```diff
@@ -0,0 +1,42 @@
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @OneToOne
+    @NotNull
```
I think this annotation is not valid for this scenario, it's from validation packet.
Here is my suggestion `@JoinColumn(name = "user_id", nullable = false)`

vadymhrnk commented on src/main/resources/db/changelog/changes/07-create_shopping_carts_table.yaml:
```diff
@@ -0,0 +1,28 @@
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
If column have `primaryKey: true`, `nullable: false` constraint can be not mentioned here

d1sam replied:
also it's up to you, but `nullable: false` can be removed :)

vadymhrnk commented on src/main/resources/db/changelog/changes/08-create_cart_items_table.yaml:
```diff
@@ -0,0 +1,40 @@
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
No new line in the end of file

vadymhrnk reviewed (changes requested):
Good job! Left for you some points to consider

DEz-1996 committed:
Some fixes:
1. Add validation to CartItemAddBookRequestDto and CartItemQuantityRequestDto, add validation in ShoppingCartController methods;
2. Change constraint annotation for field private User user in ShoppingCart
3. Delete redundant ShoppingCartUpdateRequestDto and ShoppingCartCreateRequestDto
4. Some checkstyle fixes

d1sam commented on src/main/java/online/book/store/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,71 @@
+    @Operation(
+            summary = "Retrieve user's shopping cart",
+            description = "Get info and books from cart")
+    public ShoppingCartResponseDto getCartContent(Authentication authentication) {
```
```suggestion
    public ShoppingCartResponseDto getShoppingCart(Authentication authentication) {
```

d1sam commented on src/main/java/online/book/store/repository/CartItemRepository.java:
```diff
@@ -0,0 +1,7 @@
+import online.book.store.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
```
```suggestion
@Repository
public interface CartItemRepository extends JpaRepository<CartItem, Long> {
```

d1sam commented on src/main/java/online/book/store/repository/ShoppingCartRepository.java:
```diff
@@ -0,0 +1,9 @@
+import online.book.store.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
```
```suggestion
@Repository
public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
```

d1sam commented on src/main/resources/db/changelog/changes/07-create_shopping_carts_table.yaml:
```diff
@@ -0,0 +1,28 @@
+databaseChangeLog:
+  - changeSet:
+      id: create_shopping_carts_table
+      author: DEz
```
```suggestion
      author: Volodymyr Ozemblovskyi
```
it's better to write full name of changeset's author

d1sam commented on src/main/resources/db/changelog/changes/08-create_cart_items_table.yaml:
```diff
@@ -0,0 +1,40 @@
+databaseChangeLog:
+  - changeSet:
+      id: create_cart_items_table
+      author: DEz
```
```suggestion
      author: Volodymyr Ozemblovskyi
```
it's better to write full name of changeset's author

d1sam reviewed (changes requested):
good job, but there are some minor changes should be done!

DEz-1996 committed:
Some fixes:
1. Add @Repository to CartItemRepository.java and ShoppingCartRepository.java
2. Change author name in create_shopping_carts_table.yaml and create_cart_items_table.yaml
3. Change name method from getCartContent to getShoppingCart in ShoppingCartController
